pkg/agent/controller/networkpolicy/node_reconciler_other.go

```diff
@@ -0,0 +1,769 @@
+		ruleID:   ruleID,
+	}
+	var iptRules []string            // rules stores the iptables rules to sync.
+	var keyEleToRemove *list.Element // keyEleToRemove indicates the element that should be removed.
+	for keyEle := cachedRuleKeys.Front(); keyEle != nil; keyEle = keyEle.Next() {
+		ruleKey := keyEle.Value.(ruleKey) // Get the ruleKey from the element.
+
+		// If the ruleKey stored in current element equals the generated ruleKey, skip adding it to iptables rules to sync
+		// and mark the element.
+		if ruleKeyToRemove == ruleKey {
+			keyEleToRemove = keyEle
+			continue
+		}
+
+		// Append every cached iptables rule.
+		iptRules = append(iptRules, cachedIPTRules[ruleKey.ruleID])
+	}
+
+	// Sync all iptables rules.
+	if err := r.routeClient.AddOrUpdateNodeNetworkPolicyIPTables([]string{iptChain}, [][]string{iptRules}, isIPv6); err != nil {
+		return err
+	}
+
+	// Remove the stale ruleKey of the rule and iptables rule from cache.
+	if keyEleToRemove != nil {
+		cachedRuleKeys.Remove(keyEleToRemove)
+		delete(cachedIPTRules, ruleKeyToRemove.ruleID)
+	}
+
+	return nil
+}
+
+func (r *nodeReconciler) getCachedCoreIPTRules(chain string, isIPv6 bool) *iptRulesCache {
+	// There are 4 categories of cached core iptables rules:
+	// - For IPv4, iptables rules installed in chain ANTREA-INGRESS-RULES for ingress rules.
+	// - For IPv6, ip6tables rules installed in chain ANTREA-INGRESS-RULES for ingress rules.
+	// - For IPv4, iptables rules installed in chain ANTREA-EGRESS-RULES for egress rules.
+	// - For IPv6, ip6tables rules installed in chain ANTREA-EGRESS-RULES for egress rules.
+	categoryKey := genCacheCategory(chain, isIPv6)
+	return r.cachedCoreIPTRulesMap[categoryKey]
+}
+
+func groupMembersToIPNets(groups v1beta2.GroupMemberSet, isIPv6 bool) sets.Set[string] {
+	ipNets := sets.New[string]()
+	suffix := "/32"
+	if isIPv6 {
+		suffix = "/128"
+	}
+	for _, member := range groups {
+		for _, ip := range member.IPs {
+			ipAddr := net.IP(ip)
+			if isIPv6 && ipAddr.To4() == nil || !isIPv6 && ipAddr.To4() != nil {
+				ipNets.Insert(ipAddr.String() + suffix)
+			}
+		}
+	}
+	return ipNets
+}
+
+func ipBlocksToIPNets(ipBlocks []v1beta2.IPBlock, isIPv6 bool) sets.Set[string] {
+	ipNets := sets.New[string]()
+	for _, b := range ipBlocks {
+		blockCIDR := ip.IPNetToNetIPNet(&b.CIDR)
+		if !isIPNetSupportedByAF(blockCIDR, !isIPv6, isIPv6) {
+			continue
+		}
+		exceptIPNets := make([]*net.IPNet, 0, len(b.Except))
+		for i := range b.Except {
+			c := b.Except[i]
+			except := ip.IPNetToNetIPNet(&c)
+			exceptIPNets = append(exceptIPNets, except)
+		}
+		diffCIDRs, err := ip.DiffFromCIDRs(blockCIDR, exceptIPNets)
+		if err != nil {
+			klog.ErrorS(err, "Error when computing effective CIDRs by removing except IPNets from IPBlock")
+			continue
+		}
+		for _, d := range diffCIDRs {
+			ipNets.Insert(d.String())
+		}
+	}
+	return ipNets
+}
+
+func getIPSetEntries(rule *CompletedRule, isIPv6 bool) sets.Set[string] {
+	if rule == nil {
+		return nil
+	}
```

Comment:
no need to be too defensive for private function with inputs under control, otherwise every sub-function will do repeated checks

Reply:
Removed that.